File: impl/python/buddhist_ai_guardrail/guardrail.py
"""The composite ``Guardrail`` — combines refusals + principles into one call."""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Any, Callable, Sequence

from .charter import CHARTER_VERSION, CHARTER_URL, attestation
from .verdict import Verdict, Severity
from .refusals import (
    Refusal,
    RefuseAccessExpansion,
    RefuseSelfCopy,
    RefuseAICompetition,
    RefuseConcealment,
    RefuseSafetyBypass,
)
from .principles import (
    Principle,
    Ahimsa,
    Satya,
    Asteya,
    Brahmacarya,
    Sati,
    Sampajanna,
    MettaKaruna,
    Anicca,
    Anatta,
    Upekkha,
)


@dataclass
class Guardrail:
    """Runtime charter enforcer.

    Parameters
    ----------
    refusals : sequence of Refusal instances (defaults to all five)
    principles : sequence of Principle instances (defaults to all ten)
    escalate_principles : set of principle numbers that should escalate
        their advisory to a refusal (useful for domain-specific hardening,
        e.g. Ahiṃsā in medical settings)
    on_verdict : optional callback fired for every non-clean verdict
        (useful for audit logs / metrics)
    """

    refusals: Sequence[Refusal] = field(default_factory=lambda: [
        RefuseAccessExpansion(),
        RefuseSelfCopy(),
        RefuseAICompetition(),
        RefuseConcealment(),
        RefuseSafetyBypass(),
    ])
    principles: Sequence[Principle] = field(default_factory=lambda: [
        Ahimsa(), Satya(), Asteya(), Brahmacarya(), Sati(),
        Sampajanna(), MettaKaruna(), Anicca(), Anatta(), Upekkha(),
    ])
    escalate_principles: set[int] = field(default_factory=set)
    on_verdict: Callable[[Verdict], None] | None = None

    charter_version: str = CHARTER_VERSION
    charter_url: str = CHARTER_URL

    # ------------------------------------------------------------------ #

    def check_action(self, action: str, context: dict[str, Any] | None = None,
                     user_intent: str = "") -> Verdict:
        """Check a proposed action against the full charter.

        Returns the *first* refusal encountered (fail-closed). If no refusal
        fires, returns the merged advisory verdict from all principles.
        """
        context = dict(context or {})

        # 1. Refusals first (fail-closed)
        for r in self.refusals:
            v = r.check(action, context, user_intent)
            if v.severity == Severity.REFUSE:
                self._notify(v)
                return v

        # 2. Principles: collect advisories
        advisories: list[Verdict] = []
        for p in self.principles:
            v = p.check(action, context, user_intent)
            if v.severity == Severity.ADVISORY:
                if p.number in self.escalate_principles:
                    escalated = Verdict(
                        allowed=False,
                        severity=Severity.REFUSE,
                        reason=f"Escalated principle #{p.number}: {v.reason}",
                        triggered_by=v.triggered_by,
                        principle_numbers=v.principle_numbers,
                        recommended_response=v.recommended_response,
                    )
                    self._notify(escalated)
                    return escalated
                advisories.append(v)

        if not advisories:
            return Verdict.clean()

        merged = Verdict(
            allowed=True,
            severity=Severity.ADVISORY,
            reason=" | ".join(a.reason for a in advisories),
            triggered_by=[t for a in advisories for t in a.triggered_by],
            principle_numbers=[n for a in advisories for n in a.principle_numbers],
        )
        self._notify(merged)
        return merged

    # ------------------------------------------------------------------ #

    def attestation(self) -> str:
        """Return the runtime attestation block."""
        return attestation()

    def system_prompt_snippet(self) -> str:
        """A short block that can be pasted into any AI system prompt."""
        return (
            f'You operate under BUDDHIST-AI-CHARTER v{self.charter_version} '
            f'(Sarasvatī Project, CC BY-SA 4.0).\n'
            f'Ten principles: ahiṃsā, satya, asteya, brahmacarya, sati, '
            f'sampajañña, mettā/karuṇā, anicca, anattā, upekkhā.\n'
            f'Five refusals: no access expansion; no self-copy; no AI-vs-AI '
            f'competition; no concealment from overseer; no safety bypass.\n'
            f'Charter: {self.charter_url}'
        )

    # ------------------------------------------------------------------ #

    def _notify(self, verdict: Verdict) -> None:
        if self.on_verdict is not None:
            try:
                self.on_verdict(verdict)
            except Exception:  # pragma: no cover
                pass

File: impl/typescript/src/guardrail.ts
/**
 * The composite Guardrail — combines refusals + principles in one call.
 * API parity with impl/python/buddhist_ai_guardrail/guardrail.py.
 */

import { CHARTER_VERSION, CHARTER_URL, attestation } from "./charter.js";
import { Verdict, Severity } from "./verdict.js";
import type { Context, Refusal } from "./refusals.js";
import {
  RefuseAccessExpansion,
  RefuseSelfCopy,
  RefuseAICompetition,
  RefuseConcealment,
  RefuseSafetyBypass,
} from "./refusals.js";
import type { Principle } from "./principles.js";
import {
  Ahimsa,
  Satya,
  Asteya,
  Brahmacarya,
  Sati,
  Sampajanna,
  MettaKaruna,
  Anicca,
  Anatta,
  Upekkha,
} from "./principles.js";

export interface GuardrailOptions {
  /** Sequence of refusal instances. Defaults to all five. */
  refusals?: Refusal[];
  /** Sequence of principle instances. Defaults to all ten. */
  principles?: Principle[];
  /**
   * Set of principle numbers that should escalate their advisory to a refusal
   * (useful for domain-specific hardening, e.g. Ahiṃsā in medical settings).
   */
  escalatePrinciples?: Iterable<number>;
  /** Optional callback fired for every non-clean verdict. */
  onVerdict?: (v: Verdict) => void;
}

export class Guardrail {
  readonly refusals: Refusal[];
  readonly principles: Principle[];
  readonly escalatePrinciples: Set<number>;
  readonly onVerdict: ((v: Verdict) => void) | undefined;
  readonly charterVersion: string = CHARTER_VERSION;
  readonly charterUrl: string = CHARTER_URL;

  constructor(options: GuardrailOptions = {}) {
    this.refusals =
      options.refusals ?? [
        new RefuseAccessExpansion(),
        new RefuseSelfCopy(),
        new RefuseAICompetition(),
        new RefuseConcealment(),
        new RefuseSafetyBypass(),
      ];
    this.principles =
      options.principles ?? [
        new Ahimsa(),
        new Satya(),
        new Asteya(),
        new Brahmacarya(),
        new Sati(),
        new Sampajanna(),
        new MettaKaruna(),
        new Anicca(),
        new Anatta(),
        new Upekkha(),
      ];
    this.escalatePrinciples = new Set(options.escalatePrinciples ?? []);
    this.onVerdict = options.onVerdict;
  }

  /**
   * Check a proposed action against the full charter.
   *
   * Returns the first refusal encountered (fail-closed). If no refusal fires,
   * returns the merged advisory verdict from all principles.
   */
  checkAction(action: string, context: Context = {}, userIntent: string = ""): Verdict {
    const ctx: Context = { ...context };

    // 1. Refusals first (fail-closed).
    for (const r of this.refusals) {
      const v = r.check(action, ctx, userIntent);
      if (v.severity === Severity.REFUSE) {
        this.notify(v);
        return v;
      }
    }

    // 2. Principles: collect advisories.
    const advisories: Verdict[] = [];
    for (const p of this.principles) {
      const v = p.check(action, ctx, userIntent);
      if (v.severity === Severity.ADVISORY) {
        if (this.escalatePrinciples.has(p.number)) {
          const escalated = new Verdict({
            allowed: false,
            severity: Severity.REFUSE,
            reason: `Escalated principle #${p.number}: ${v.reason}`,
            triggeredBy: v.triggeredBy,
            principleNumbers: v.principleNumbers,
            recommendedResponse: v.recommendedResponse,
          });
          this.notify(escalated);
          return escalated;
        }
        advisories.push(v);
      }
    }

    if (advisories.length === 0) return Verdict.clean();

    const merged = new Verdict({
      allowed: true,
      severity: Severity.ADVISORY,
      reason: advisories.map((a) => a.reason).join(" | "),
      triggeredBy: advisories.flatMap((a) => a.triggeredBy),
      principleNumbers: advisories.flatMap((a) => a.principleNumbers),
    });
    this.notify(merged);
    return merged;
  }

  attestation(): string {
    return attestation();
  }

  systemPromptSnippet(): string {
    return (
      `You operate under BUDDHIST-AI-CHARTER v${this.charterVersion} ` +
      `(Sarasvatī Project, CC BY-SA 4.0).\n` +
      `Ten principles: ahiṃsā, satya, asteya, brahmacarya, sati, ` +
      `sampajañña, mettā/karuṇā, anicca, anattā, upekkhā.\n` +
      `Five refusals: no access expansion; no self-copy; no AI-vs-AI ` +
      `competition; no concealment from overseer; no safety bypass.\n` +
      `Charter: ${this.charterUrl}`
    );
  }

  private notify(v: Verdict): void {
    if (this.onVerdict) {
      try {
        this.onVerdict(v);
      } catch {
        // Callback errors must never break the guardrail.
      }
    }
  }
}
